rpcs3/Emu/Cell/lv2/sys_uart.cpp

```diff
@@ -1,35 +1,597 @@
+		}
+
+		u32 read_size;
+
+		for (read_size = ITER_SIZE; read_size + ITER_SIZE < buf_size; read_size += ITER_SIZE)
+		{
+			const u32 in_loop_read = vuart_thread.read_rx_data(buf + read_size, ITER_SIZE);
+
+			if (in_loop_read < ITER_SIZE)
+			{
+				return read_size + in_loop_read;
+			}
+		}
+
+		return read_size + vuart_thread.read_rx_data(buf + read_size, buf_size - read_size);
+	};
+
+	if (mode == 1)
+	{
+		// Yield before checking for packets
+		lv2_obj::sleep(ppu);
+
+		for (;;)
+		{
+			if (ppu.is_stopped())
+			{
+				return {};
+			}
+
+			std::unique_lock<shared_mutex> lock(vuart_thread.rx_wake_m);
+			const s32 read_result = vuart_read(data.get(), static_cast<u32>(size));
+
+			if (read_result > CELL_OK)
+			{
+				read_size = read_result;
+				break;
+			}
+
+			vuart_thread.rx_wake_c.wait_unlock(5000, lock);
+		}
+
+		ppu.check_state();
+	}
+	else
+	{
+		const s32 read_result = vuart_read(data.get(), static_cast<u32>(size));
+
+		if (read_result <= CELL_OK)
+		{
+			return read_result;
+		}
+
+		read_size = read_result;
+	}
+
+	if (!buffer || !vm::check_addr(buffer.addr(), vm::page_writable, read_size))
+	{
+		return CELL_EFAULT;
+	}
+
+	memcpy(buffer.get_ptr(), data.get(), read_size);
+	return not_an_error(read_size);
 }
 
-error_code sys_uart_send(vm::cptr<void> buffer, u64 size, u64 flags)
+error_code sys_uart_send(ppu_thread &ppu, vm::cptr<void> buffer, u64 size, u32 mode)
 {
-	sys_uart.todo("sys_uart_send(buffer=0x%x, size=0x%llx, flags=0x%x)", buffer, size, flags);
+	sys_uart.trace("sys_uart_send(buffer=0x%x, size=0x%llx, mode=0x%x)", buffer, size, mode);
 
-	return CELL_OK;
+	if (!size)
+	{
+		return CELL_OK;
+	}
+
+	if ((mode & ~3ul) != 0)
+	{
+		return CELL_EINVAL;
+	}
+
+	auto &vuart_thread = g_fxo->get<vuart_av>();
+
+	if (!vuart_thread.init)
+	{
+		return CELL_ESRCH;
+	}
+
+	if (size > 0x20000U)
+	{
+		// kmalloc restriction
+		fmt::throw_exception("Buffer is too big");
+	}
+
+	if (!vm::check_addr(buffer.addr(), vm::page_readable, static_cast<u32>(size)))
+	{
+		return CELL_EFAULT;
+	}
+
+	const std::unique_ptr<u8[]> data = std::make_unique<u8[]>(size);
+	memcpy(data.get(), buffer.get_ptr(), size);
+
+	std::unique_lock lock(vuart_thread.tx_mutex, std::defer_lock);
+
+	const u32 ITER_SIZE = 4096;
+
+	if (mode & 1)
+	{
+		// Yield before sending packets
+		lv2_obj::sleep(ppu);
+
+		lock.lock();
+
+		auto vuart_send_all = [&](const u8 *data, u32 data_sz)
+		{
+			u32 rem_size = data_sz;
+
+			while (rem_size)
+			{
+				if (ppu.is_stopped())
+				{
+					return false;
+				}
+
+				std::unique_lock<shared_mutex> lock(vuart_thread.tx_rdy_m);
+				if (vuart_thread.get_tx_bytes() >= PS3AV_TX_BUF_SIZE)
+				{
+					vuart_thread.tx_rdy_c.wait_unlock(5000, lock);
+				}
+				else
+				{
+					lock.unlock();
+				}
+				rem_size -= vuart_thread.enque_tx_data(data + data_sz - rem_size, rem_size);
+			}
+
+			return true;
+		};
+
+		if (!vuart_send_all(data.get(), std::min(ITER_SIZE, static_cast<u32>(size)))) return {};
```

Comment:
This section and the one below can be combined into one loop from what I can see, with far less complexity. It can easily be one for lopp with only one line containing vuart_send_all(...)

Reply:
See other comment for suggested simplification.

Reply:
Done